### PR TITLE
Alerting: Export and provisioning rules into subfolders

### PR DESCRIPTION
**Pull request by papagian**
**What is this feature?**

In this work we change to use [the folder fulllpath instead of the folder title](https://docs.google.com/document/d/1XdtwT_Ke2M_492DsL8G89JccilcovHguH4H0CJTAIx0/edit#heading=h.e2bjknto5jx6) when exporting and provisioning alert rules.

**Why do we need this feature?**

Currently provisioning operates on the assumption the folder tile is guaranteed to be unique throughout the organization. However, with nested folders introduced, after #77076 is merged, this won't be the case.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**
To test:
1. start grafana server in `main`
2. create two folders: one containing slash in its title
3. create 2 alerts under the above folders
4. export the alert rules using the [GET /api/v1/provisioning/alert-rules](https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/#route-get-alert-rules) API; note that slashes in the  folder title **are not escaped**
5. create another org
6. change the `orgId` in the exported file to point to the new org and save it under `conf/provisioniong/alerting/alerts.yaml`; the provisioning file should look like the following (the identifiers will different):

<details><summary>example provisioning file</summary>
<p>

```yaml
apiVersion: 1
groups:
    - orgId: 2
      name: 5m
      folder: parent
      interval: 5m
      rules:
        - uid: b47d7786-3d17-42e3-b164-9960e58f0f94
          title: zebra 1
          condition: C
          data:
            - refId: A
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: PD8C576611E62080A
              model:
                intervalMs: 1000
                maxDataPoints: 43200
                refId: A
            - refId: B
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params: []
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - B
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: A
                intervalMs: 1000
                maxDataPoints: 43200
                reducer: last
                refId: B
                type: reduce
            - refId: C
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params:
                            - 0
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - C
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: B
                intervalMs: 1000
                maxDataPoints: 43200
                refId: C
                type: threshold
          dashboardUid: ""
          panelId: 0
          noDataState: NoData
          execErrState: Error
          for: 5m
          annotations: {}
          labels: {}
          isPaused: false
    - orgId: 2
      name: 5m
      folder: folder_with_/in_name
      interval: 5m
      rules:
        - uid: c5a1312f-d0b0-4d8d-bb76-99a35bbf9079
          title: rule under child
          condition: C
          data:
            - refId: A
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: PD8C576611E62080A
              model:
                intervalMs: 1000
                maxDataPoints: 43200
                refId: A
                scenarioId: csv_metric_values
                stringInput: 1,20,90,30,5,0
            - refId: B
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params: []
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - B
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: A
                intervalMs: 1000
                maxDataPoints: 43200
                reducer: last
                refId: B
                type: reduce
            - refId: C
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params:
                            - 0
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - C
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: B
                intervalMs: 1000
                maxDataPoints: 43200
                refId: C
                type: threshold
          dashboardUid: ""
          panelId: 0
          noDataState: NoData
          execErrState: Error
          for: 5m
          annotations: {}
          labels: {}
          isPaused: false
```
</p>
</details> 

7. restart the server and check that the folders and the alerts are provisioned under the new organization (created in step 5)
Screenshot 2023-11-17 at 10 54 27 AM
Screenshot 2023-11-17 at 10 54 39 AM

8. switch to this branch
9. escape slash character from the folder title in the above provisioning file
10. add a new entry for provisioning an alert in a subfolder like the following:

<details><summary>example additional entry provisioning file</summary>
<p>

```yaml
apiVersion: 1
groups:
    - orgId: 2
      name: 5m
      folder: parent
      interval: 5m
      rules:
        - uid: b47d7786-3d17-42e3-b164-9960e58f0f94
          title: zebra 1
          condition: C
          data:
            - refId: A
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: PD8C576611E62080A
              model:
                intervalMs: 1000
                maxDataPoints: 43200
                refId: A
            - refId: B
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params: []
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - B
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: A
                intervalMs: 1000
                maxDataPoints: 43200
                reducer: last
                refId: B
                type: reduce
            - refId: C
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params:
                            - 0
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - C
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: B
                intervalMs: 1000
                maxDataPoints: 43200
                refId: C
                type: threshold
          dashboardUid: ""
          panelId: 0
          noDataState: NoData
          execErrState: Error
          for: 5m
          annotations: {}
          labels: {}
          isPaused: false
    - orgId: 2
      name: 5m
      folder: folder_with_\/in_name
      interval: 5m
      rules:
        - uid: c5a1312f-d0b0-4d8d-bb76-99a35bbf9079
          title: rule under child
          condition: C
          data:
            - refId: A
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: PD8C576611E62080A
              model:
                intervalMs: 1000
                maxDataPoints: 43200
                refId: A
                scenarioId: csv_metric_values
                stringInput: 1,20,90,30,5,0
            - refId: B
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params: []
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - B
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: A
                intervalMs: 1000
                maxDataPoints: 43200
                reducer: last
                refId: B
                type: reduce
            - refId: C
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params:
                            - 0
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - C
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: B
                intervalMs: 1000
                maxDataPoints: 43200
                refId: C
                type: threshold
          dashboardUid: ""
          panelId: 0
          noDataState: NoData
          execErrState: Error
          for: 5m
          annotations: {}
          labels: {}
          isPaused: false
    - orgId: 2
      name: 5m
      folder: folder_with_\/in_name/and_child_with_\/in_name
      interval: 5m
      rules:
        - uid: uid
          title: rule under child
          condition: C
          data:
            - refId: A
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: PD8C576611E62080A
              model:
                intervalMs: 1000
                maxDataPoints: 43200
                refId: A
                scenarioId: csv_metric_values
                stringInput: 1,20,90,30,5,0
            - refId: B
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params: []
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - B
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: A
                intervalMs: 1000
                maxDataPoints: 43200
                reducer: last
                refId: B
                type: reduce
            - refId: C
              relativeTimeRange:
                from: 600
                to: 0
              datasourceUid: __expr__
              model:
                conditions:
                    - evaluator:
                        params:
                            - 0
                        type: gt
                      operator:
                        type: and
                      query:
                        params:
                            - C
                      reducer:
                        params: []
                        type: last
                      type: query
                datasource:
                    type: __expr__
                    uid: __expr__
                expression: B
                intervalMs: 1000
                maxDataPoints: 43200
                refId: C
                type: threshold
          dashboardUid: ""
          panelId: 0
          noDataState: NoData
          execErrState: Error
          for: 5m
          annotations: {}
          labels: {}
          isPaused: false
```
</p>
</details> 

11. restart server with nested folder enabled and make sure that the new subfolder and alert are provisioned:

Screenshot 2023-11-17 at 11 04 06 AM
Screenshot 2023-11-17 at 11 06 11 AM

12. export the alert rules using the [GET /api/v1/provisioning/alert-rules](https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/#route-get-alert-rules) API; note that slashes in the  folder titles **are escaped**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Release notice breaking change

Users that provision alert rules into folders whose titles contain slashes from now on they should escape them:
eg. if an alert group contains:
`folder: folder_with_/_in_title`
it should become:
`folder: folder_with_\/_in_title`

### CODE REVIEW

File: pkg/services/dashboards/service/dashboard_service.go [hunk 1]
```diff
@@ -324,42 +324,16 @@ func (dr *DashboardServiceImpl) SaveProvisionedDashboard(ctx context.Context, dt
...
-		alerts, err := dr.dashAlertExtractor.GetAlerts(ctx, dashAlertInfo)
-		if err != nil {
-			return nil, err
-		}
-
-		err = dr.dashboardStore.SaveAlerts(ctx, dash.ID, alerts)
-		if err != nil {
-			return nil, err
-		}
-	}
```
Comment (papagian) [inline pkg/services/dashboards/service/dashboard_service.go:356]: no need to check for legacy alerts in a folder (this is meaningful only for dashboards)

File: pkg/services/dashboards/service/dashboard_service.go [hunk 2]
```diff
@@ -324,42 +324,16 @@ func (dr *DashboardServiceImpl) SaveProvisionedDashboard(ctx context.Context, dt
...
-
-	if dto.Dashboard.ID == 0 {
-		dr.setDefaultPermissions(ctx, dto, dash, true)
-	}
-
-	return dash, nil
+	dr.setDefaultFolderPermissions(ctx, dto, f, true)
```
Comment (papagian) [inline pkg/services/dashboards/service/dashboard_service.go:335]: use separate function for setting folder permissions

**Review by suntala — COMMENTED** [review]
Thanks for the detailed instructions for testing! They look very helpful. I am trying to go through them to test the change, but I'm having trouble adding alerting rules for some reason while in `main`. Something about:

```
Failed to build evaluator for queries and expressions: expression 'A' 
cannot reference itself. Must be query or another expression
```

~~I'm going to have to figure out a fix for that before looking through the rest of the instructions.~~ --> discussed on Slack.

As mentioned, please let me know if you haven't gotten a review by the time I’m back.

**Comment by undef1nd** [comment]
Hello! I've tested it and slashes escaping works as expected. But there's one other thing - created folders are not stored in `folders` table thus a request to folders service returns 404. I'm gonna investigate a bit why this is happening.

folderSvcResp

**Review by yuri-tceretian — COMMENTED** [review]
LGTM. Changes work as expected. Some comments below would be great to address but are not blockers.

However, I noticed two things that worry me:
1. I can create a folder `/`. Perhaps we should not let users do that.
2. When I provision rules and specify folder `TEST/Sub\/test/Alerting Folder/DAta\/TEST///` (3 slashes at the end), it ignores them. I wonder if we should just fail to provision in such cases (when subfolder title is empty)

**Review by yuri-tceretian — APPROVED** [review]

**Pull request merged**